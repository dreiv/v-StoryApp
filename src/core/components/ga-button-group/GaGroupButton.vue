<script setup lang="ts">
import { computed, inject, useCssModule, useTemplateRef } from 'vue'
import { buttonGroupKey } from './types'

export interface SegmentButtonProps {
  value?: string | number
  label?: string
  icon?: boolean
  disabled?: boolean
}
const { icon, value, disabled, label } = defineProps<SegmentButtonProps>()
const inputRef = useTemplateRef<HTMLElement>('inputRef')

const group = inject(buttonGroupKey)
if (!group) throw new Error('GaSegmentButton must be used inside a GaButtonGroup')

const style = useCssModule()
const classes = computed(() => [style.button, { [style.icon]: icon }])

defineExpose({ inputRef })
</script>

<template>
  <label :class="classes">
    <input
      ref="inputRef"
      type="radio"
      :class="$style.native"
      :disabled="disabled"
      :name="group.name"
      :checked="group.model.value === value"
      @change="group.model.value = value as string"
      v-bind="$attrs"
    />

    <slot>{{ label }}</slot>
  </label>
</template>

<style module>
.button {
  display: inline-flex;
  position: relative;
  align-items: center;
  gap: var(--ga-size-spacing-03);
  border: var(--ga-size-border-width-sm) solid transparent;

  border-radius: var(--ga-radius);
  padding-inline: var(--ga-size-spacing-05);
  min-width: 2.25rem; /* TODO: fix */
  height: 2.25rem; /* TODO: fix */
  user-select: none;

  &:has(input:enabled) {
    cursor: pointer;
  }

  &:has(input:focus-visible) {
    outline: var(--ga-size-border-width-md) solid var(--ga-color-border-focus);
    outline-offset: var(--ga-size-spacing-01);
  }

  &:has(input:checked) {
    border-color: var(--ga-color-border-focus);
    background-color: var(--ga-color-surface-primary);
  }

  &:has(input:not(:active, :disabled):is(:hover, :focus)) {
    border-color: var(--ga-color-border-focus);
    background-color: var(--ga-color-border-action-hover-2);
  }

  &:has(input:disabled) {
    cursor: not-allowed;
    color: var(--ga-color-text-disabled);
  }
}

.native {
  position: absolute;
  opacity: 0;
  width: 0;
  height: 0;
}

.icon {
  justify-content: center;
  padding: 0;

  &:has(input:not(:active, :disabled):is(:hover)) {
    color: var(--ga-color-icon-action-hover);
  }
}
</style>
