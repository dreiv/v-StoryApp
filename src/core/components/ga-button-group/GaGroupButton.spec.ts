import { shallowMount, type VueWrapper } from '@vue/test-utils'
import GaGroupButton from './GaGroupButton.vue'

vi.mock('./types', () => ({ buttonGroupKey: 'buttonGroup' }))

const buttonGroup = { name: 'mockName', model: { value: '' } }
describe('GroupButton', () => {
  let wrapper: VueWrapper

  beforeEach(() => {
    wrapper = shallowMount(GaGroupButton, { global: { provide: { buttonGroup } } })
  })

  it('should render correctly with default props', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render correctly with label prop', () => {
    wrapper = shallowMount(GaGroupButton, {
      props: { label: 'Button option' },
      global: { provide: { buttonGroup } },
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.text()).toContain('Button option')
  })

  it('should render correctly with default slot', () => {
    wrapper = shallowMount(GaGroupButton, {
      slots: { default: 'Custom Button Content' },
      global: { provide: { buttonGroup } },
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.text()).toContain('Custom Button Content')
  })

  it('should render with icon style when icon prop is true', async () => {
    await wrapper.setProps({ icon: true })

    const label = wrapper.find('label')
    expect(label.classes()).toContain('icon')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should be checked when group modelValue matches button value', () => {
    wrapper = shallowMount(GaGroupButton, {
      attrs: { value: 'option1' },
      global: { provide: { buttonGroup: { name: 'mockName', model: { value: 'option1' } } } },
    })

    const input = wrapper.find('input')
    expect(input.element.checked).toBe(true)
  })

  it('should emit change to update group modelValue when clicked', async () => {
    const mockModel = { value: '' }
    const buttonGroup = { name: 'mockName', model: mockModel }

    wrapper = shallowMount(GaGroupButton, {
      attrs: { value: 'newValue' },
      global: { provide: { buttonGroup } },
    })

    await wrapper.find('input').setValue(true)
    expect(mockModel.value).toBe('newValue')
  })

  it('should be disabled when disabled prop is true', async () => {
    await wrapper.setProps({ disabled: true })

    const input = wrapper.find('input')
    expect(input.element.disabled).toBe(true)
  })

  it('should throw an error when not used inside a GaButtonGroup', () => {
    expect(() => {
      shallowMount(GaGroupButton)
    }).toThrow('GaSegmentButton must be used inside a GaButtonGroup')
  })
})
