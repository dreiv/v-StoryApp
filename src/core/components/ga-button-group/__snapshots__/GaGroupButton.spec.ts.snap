// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GroupButton > should render correctly with default props 1`] = `"<label class="button"><input type="radio" class="native" name="mockName"></label>"`;

exports[`GroupButton > should render correctly with default slot 1`] = `"<label class="button"><input type="radio" class="native" name="mockName">Custom Button Content</label>"`;

exports[`GroupButton > should render correctly with label prop 1`] = `"<label class="button"><input type="radio" class="native" name="mockName">Button option</label>"`;

exports[`GroupButton > should render with icon style when icon prop is true 1`] = `"<label class="button icon"><input type="radio" class="native" name="mockName"></label>"`;
